src/daemon/default_vm_image_vault.cpp

```diff
@@ -438,36 +454,47 @@ mp::VMImage mp::DefaultVMImageVault::fetch_image(const FetchType& fetch_type, co
         }
         DeleteOnException image_file{source_image.image_path};
 
-        url_downloader->download_to(info.image_location, source_image.image_path, info.size, LaunchProgress::IMAGE,
-                                    monitor);
+        auto future = QtConcurrent::run([&]() -> VMImage {
+            try
+            {
+                url_downloader->download_to(info.image_location, source_image.image_path, info.size,
+                                            LaunchProgress::IMAGE, monitor);
 
-        monitor(LaunchProgress::VERIFY, -1);
-        verify_image_download(source_image.image_path, id);
+                monitor(LaunchProgress::VERIFY, -1);
+                verify_image_download(source_image.image_path, id);
 
-        if (fetch_type == FetchType::ImageKernelAndInitrd)
-        {
-            source_image = fetch_kernel_and_initrd(info, source_image, image_dir, monitor);
-        }
+                if (fetch_type == FetchType::ImageKernelAndInitrd)
+                {
+                    source_image = fetch_kernel_and_initrd(info, source_image, image_dir, monitor);
+                }
 
-        if (source_image.image_path.endsWith(".xz"))
-        {
-            source_image = extract_downloaded_image(source_image, monitor);
-        }
+                if (source_image.image_path.endsWith(".xz"))
+                {
+                    source_image = extract_downloaded_image(source_image, monitor);
+                }
 
-        auto prepared_image = prepare(source_image);
-        prepared_image_records[id] = {prepared_image, query, std::chrono::system_clock::now()};
-        remove_source_images(source_image, prepared_image);
+                auto prepared_image = prepare(source_image);
+                remove_source_images(source_image, prepared_image);
+
+                return prepared_image;
+            }
+            catch (const std::exception& e)
+            {
+                throw CreateImageException(e.what());
+            }
+        });
 
-        VMImage vm_image;
+        in_progress_image_fetches[id] = future;
+        lock.unlock();
 
-        if (!query.name.empty())
-        {
-            vm_image = image_instance_from(query.name, prepared_image);
-            instance_image_records[query.name] = {vm_image, query, std::chrono::system_clock::now()};
-        }
+        auto prepared_image = future.result();
 
-        persist_image_records();
-        persist_instance_records();
+        lock.lock();
+        prepared_image_records[id] = {prepared_image, query, std::chrono::system_clock::now()};
+        auto vm_image = finalize_image_records(query, prepared_image);
+
+        in_progress_image_fetches.erase(id);
+        lock.unlock();
```

Comment:
I am having a lot of trouble reading this and following all the locking and unlocking. Why is it needed? 

Repeated locking and unlocking means one thing: identifiable work units that could be extracted out. I think this would really profit from a different sort of organization, extracting individual operations and having them lock internally where needed. Something like:

```c++
QFuture<VMImage> get_existing_or_create_future_vm(const std::string& id)
{
    std::unique_lock<decltype(fetch_mutex)> lock{fetch_mutex};
    //...
    return future;
}

VMImage DefaultImageVault::fetch_image(...)
{
    auto future = get_existing_or_create_future_vm(id);
    auto my_image = future.result();
    further_operation(my_image); // if this needs to lock, it does inside
    // ...
    return my_image;
}
```

Reply:
The locking an unlocking is needed so different threads don't read/write to the json files that is done in `finalize_image_records` at the same time and potentially corrupt the files if writing at the same time.

I agree the locking/unlocking is messy and organizing and breaking down this function would be beneficial- we just need to weigh the time vs. benefit of doing such a thing.

Reply:
IMO it is bound to save us time in the ~~long~~ medium-run.

Reply:
Ok, I'm looking at this and I think I can do better, especially around using `lock_guard`'s.

Reply:
As mentioned above, I used `lock_guard`'s everywhere to take advantage of scoped based locking.  Hopefully this addresses some of your concerns.